### ISSUE TITLE
[Feature request]: Support exporting scripts

### ISSUE THREAD

**Issue by matthiasclasen** [flatpak/flatpak#5395]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

In https://github.com/flatpak/xdg-desktop-portal/pull/705, we are working on supporting the web browsers 'native messaging host' mechanism in flatpaks, via  a portal. In order to support using flatpaked apps as messaging hosts, it would be very useful if flatpak allowed exporting script wrappers that are a bit more flexible than the generated ones we have now.

See https://github.com/flatpak/xdg-desktop-portal/pull/705#discussion_r1028074268 for a detailed discussion.

**Comment by ramcq** [comment]
My idea was that in addition to `command: blah` in the manifest, we also add `commands: ['blah', 'blah', 'blah']` so that additional executable entry points can be added to Flatpaks. Then in addition to the app ID wrapper scripts written out here https://github.com/flatpak/flatpak/blob/main/common/flatpak-dir.c#L8944 we would also write `app.id-blah` wrappers with each of the commands.

Then we are able to add the native messaging host path to the list at https://github.com/flatpak/flatpak/blob/main/common/flatpak-dir.c#L7984 and - provided the binary name is one of the commands above - substitute the executable name for the wrapper script which is added above. Note that the standard does not allow passing arguments to these native host binaries, which is why we must go via the wrapper script. Personally I believe that having a general-purpose multiple binary export mechanism is more useful for other cases than having a special-case just for native messaging hosts.

Unsolved so far in this plan is how a) the browsers discover the exported native messaging hosts which will be in a non-standard path (on XDG_DATA_DIRS, but not necessarily checked by the browsers?), and b) for a browser running inside a flatpak, how the JSON files exported become visible or mapped back in to a path that the browser can see.

**Comment by smcv** [comment]
> Then in addition to the app ID wrapper scripts written out here https://github.com/flatpak/flatpak/blob/main/common/flatpak-dir.c#L8944 we would also write `app.id-blah` wrappers with each of the commands.

If I remember correctly we allow dashes in the last dot-separated segment of an app ID, so they might have to be `app.id.blah`.

> Personally I believe that having a general-purpose multiple binary export mechanism is more useful for other cases than having a special-case just for native messaging hosts.

A good concrete use-case for this might be that `com.valvesoftware.Steam` contains both `steam-wrapper` and `steamcmd-wrapper` entry points.

Are you saying that `com.valvesoftware.Steam` should additionally export `.../flatpak/exports/bin/com.valvesoftware.Steam.steamcmd` (or `...Steam-steamcmd` in your proposed naming), which would have behaviour equivalent to `flatpak run --command=steamcmd-wrapper com.valvesoftware.Steam`?

Maybe it should also export `.../flatpak/exports/bin/com.valvesoftware.Steam.steam`, which would be equivalent to `flatpak run com.valvesoftware.Steam` with the default command `steam-wrapper`, but more explicit?

**Comment by smcv** [comment]
Or, they could be named like `com.valvesoftware.Steam+steamcmd` using a separator that isn't allowed in app-IDs (for which I think `+`, `:`, `,` and maybe `=` and `@` are the only options that aren't going to be annoying to use from a shell).

**Comment by ramcq** [comment]
> A good concrete use-case for this might be that `com.valvesoftware.Steam` contains both `steam-wrapper` and `steamcmd-wrapper` entry points.
> 
> Are you saying that `com.valvesoftware.Steam` should additionally export `.../flatpak/exports/bin/com.valvesoftware.Steam.steamcmd` (or `...Steam-steamcmd` in your proposed naming), which would have behaviour equivalent to `flatpak run --command=steamcmd-wrapper com.valvesoftware.Steam`?

:+1: 

> Maybe it should also export `.../flatpak/exports/bin/com.valvesoftware.Steam.steam`, which would be equivalent to `flatpak run com.valvesoftware.Steam` with the default command `steam-wrapper`, but more explicit?

I'm relaxed on whether we effectively prepend `command` at `commands[0]`. It does have the advantage of if you are trying to autocomplete shell commands from the exported wrapper scripts, the "obvious" one is also always there.

> Or, they could be named like `com.valvesoftware.Steam+steamcmd` using a separator that isn't allowed in app-IDs (for which I think `+`, `:`, `,` and maybe `=` and `@` are the only options that aren't going to be annoying to use from a shell).

Yeah maybe a + is safer to avoid aliasing. You just know someone will export a command named `desktop` and the universe will implode. :rofl:

**Comment by ramcq** [comment]
Hmm... I'm a little skeptical here about whether we really need the renaming. Mostly because it causes me to go back and question the behaviour of automatically exporting the --command, because it doesn't have the opportunity to be renamed. Inspecting the bin/ dirs of the Flatpaks I have installed, overwhelmingly the binaries are named "normally" in bin, and even those with wrappers manage to have the binary renamed and the wrapper has the expected name and the underlying binary is named or located differently (be it foo.bin, or /app/extra/..., or /app/appname/appbin, etc).

**Comment by smcv** [comment]
Yes, I can see an argument that the Steam Flatpak should rename the underlying executables provided by Valve to `steam.real` and `steamcmd.real` or something (or move them to a directory off the `PATH`, or similar), and then rename its `steam-wrapper` and `steamcmd-wrapper` to `steam` and `steamcmd` (possibly with a compatibility symlink).
